spec/models/provider_spec.rb

```diff
@@ -61,4 +61,32 @@
       expect(tenant.providers).to include(provider)
     end
   end
+
+  context "#destroy_queue" do
+    before do
+      allow(Time).to receive(:now).and_return(Time.zone.now)
+    end
+    it "destroy when has no managers" do
+      expect(provider).to receive(:destroy)
+      provider.destroy_queue
+    end
+
+    it "to destroy_queue its managers first" do
+      manager = FactoryGirl.create(:ext_management_system, :enabled => true)
+      provider.managers = [manager]
+      expect(manager).to receive(:destroy_queue)
+      expect(provider).not_to receive(:destroy)
+      expect(described_class).to receive(:_queue_task).with(:destroy_queue, provider.id.to_miq_a, 15.seconds.from_now)
+      provider.destroy_queue
+    end
+
+    it "doesn't destroy_queue its managers when they are disabled" do
+      manager = FactoryGirl.create(:ext_management_system, :enabled => false)
+      provider.managers = [manager]
+      expect(manager).not_to receive(:destroy_queue)
+      expect(provider).not_to receive(:destroy)
```

Comment:
Won't this case sit in a loop forever?

Reply:
This is the case when managers are already in the process of being destroyed. Manager are being disabled first to signal to their workers to go down. Refer to [here](https://github.com/jameswnl/manageiq/blob/f62dae6ec44b686d1842d117248e28d818c6b69a/app/models/ext_management_system.rb#L462-L462)

Reply:
I feel like there are edge cases that aren't covered here. (A provider added but it's disabled for some reason)